### PR TITLE
Add endpoint authentication

### PR DESCRIPTION
**Pull request by zachlipton**
- Require a bearer token for endpoints that create/update data
- Add explanation to the README
- Add a `flask auth getToken` CLI command to obtain a token
- Add a test fixture to make authenticated calls in tests

The additions to the README pretty much explain how this works practically. A few notes:

- There's a matching component for this to send an authentication token from the spreadsheet
- I went ahead and used [JWT Tokens](https://jwt.io/introduction/) despite us not really taking full advantage of them at this point. For the moment, any valid token signed against the secret key will be accepted, and we're not differentiating levels of access or other payload data. However, this design will allow us in the future to have goodies like tokens with different privilege levels and a revoked tokens list.
- Currently, tokens never expire. This relies on our ability to not leak the token in the Apps Script's config file. I believe only authorized team members can view the Apps Script on the sheet, but this needs to be confirmed.
- All valid tokens can be revoked by changing the secret key
- This is broadly tested (things that are supposed to require a valid token fail if one isn't supplied), but not rigorously audited from a security standpoint
- [ ] When this is deployed to a server, the `SECRET_KEY` environment variable must be configured with a secret or nothing will be secure

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -19,6 +20,10 @@ def create_app(config):
 
     db.init_app(app)
     migrate.init_app(app, db)
+    
+    # setup flask_jwt_extended for authentication
+    app.config['JWT_SECRET_KEY'] = app.config['SECRET_KEY']
```
Comment (julia326) [inline app/__init__.py:25]: should we error with some helpful message if `SECRET_KEY` doesn't exist in the config, or at least print out an angry message if the value is the default `12345`?

File: README.md [hunk 2]
```diff
@@ -58,6 +58,25 @@ You can run the tests with `python -m pytest` once your environemnt is configure
...
+## Authentication
+
+Endpoints that create/update data are authenticated with a JWT bearer token. 
+
+To obtain a token, run:
+```shell
+flask auth getToken tokenName
+```
+
+The token is secured by the value of `FLASK_SECRET`, so tokens generated in one environment will not work elsewhere
```
Comment (julia326) [inline README.md:70]: do you mean `SECRET_KEY`?

File: config.py [hunk 3]
```diff
@@ -21,6 +21,8 @@ def SQLALCHEMY_DATABASE_URI(self):
     SQLALCHEMY_RECORD_QUERIES = True
 
     SECRET_KEY = env_conf("SECRET_KEY", cast=str, default="12345")
+    # by default, access tokens do not expire
+    JWT_ACCESS_TOKEN_EXPIRES = env_conf('JWT_ACCESS_TOKEN_EXPIRES', cast=int, default=False)
```
Comment (julia326) [inline config.py:25]: cast=bool?

File: tests/app/v1_basic_test.py [hunk 4]
```diff
@@ -19,7 +19,7 @@ def test_get_test(app):
     assert "test_data_key" in data 
     assert data["test_data_key"] == "test_data_value" 
 
-def test_post_core_data(app):
+def test_post_core_data(app, headers):
```
Comment (julia326) [inline tests/app/v1_basic_test.py:22]: can we also add a small test that a POST without authentication will fail?

File: app/__init__.py (hunk 1 above)
Comment (zachlipton) [reply app/__init__.py:25]: Good idea. I added an assertion that will make the server fail to launch if the default `SECRET_KEY` is used. It doesn't fire in the development or testing environments so it shouldn't annoy us when we're working locally.

File: README.md (hunk 2 above)
Comment (zachlipton) [reply README.md]: Yes, but think how much fun I could have had spending an hour swearing at it when I used the wrong environment variable name on the server and couldn't figure out what was wrong

File: config.py (hunk 3 above)
Comment (zachlipton) [reply config.py:25]: It seems like it, but it's normally an integer value that corresponds to how long the token should be valid. False is just an override for "never expire."

File: tests/app/v1_basic_test.py (hunk 4 above)
Comment (zachlipton) [reply tests/app/v1_basic_test.py:22]: Good idea. I had a general test for that in auth_test but I added a test specifically for this endpoint

**Pull request merged**